products/pdt/pdtlisten/src/listener.rs

```diff
@@ -0,0 +1,87 @@
+use ethers::{
+    providers::{Http, Middleware, Provider},
+    types::{Block, Transaction, U64},
+};
+
+use anyhow::Result;
+use async_stream::try_stream;
+use serde_json::{to_value, Value};
+use tokio::time::{interval, Duration};
+use tokio_stream::Stream;
+
+async fn get_block_by_number(x: U64, provider: &Provider<Http>) -> Result<Block<Transaction>> {
+    println!("found block with number {:?}, getting block info", x);
+    fn serialize(v: impl serde::Serialize) -> Value {
+        to_value(v).unwrap()
+    }
+    let mut raw_block: Value = provider
+        .request("eth_getBlockByNumber", [serialize(x), serialize(true)])
+        .await?;
+    // ZIL-5328 means our nonce is only one byte instead of 8, which ethers
+    // is not happy about.
+    raw_block["nonce"] = serde_json::to_value("0x0000000000000000")?;
+
+    let mut block: Block<Transaction> = serde_json::from_value(raw_block)?;
+    while block.number.is_none() {
```

Comment:
Have you seen responses where `block.number.is_none()`? I would be surprised if it was possible.

Reply:
It does happen sometimes when a block that is retrieved is somehow still pending. I have seen it whilst I was testing. Hence I maintained this code to the same as `get_block_by_hash`